src/dune_rules/action_unexpanded.ml

```diff
@@ -479,8 +551,22 @@ let rec expand (t : Dune_lang.Action.t) ~context : Action.t Action_expander.t =
   | Cram script ->
     let+ script = E.dep script in
     Cram_exec.action script
+  | Ppx ({ pps; flags; loc }, input_file) -> (
+    match ml_kind with
+    | None ->
+      Code_error.raise ~loc
```

Comment:
I wonder if this is is "code error" or "user error". For example, this stanza triggers this case:

```
(rule
 (action
  (progn
   (run cat %{input-file})
   (ppx (some_ppx) %{input-file}.pp.1))))
```

It would be more informative to tell users that this case is not supported, rather than show stack trace and "Uncertainty is the mind-killer" error message.

Alternatively, can't dune just forbid the usage of `ppx` in simple user rules? Is there a use case for that?

Reply:
>  Is there a use case for that?

It can be useful to write expect tests for ppx

Reply:
I'll add a cram test for this error that will be fixed in a future PR.

Reply:
added a cram test in [test/blackbox-tests/test-cases/preprocess-with-action-and-pps-errors.t](https://github.com/ocaml/dune/pull/7373/files#diff-068ca8d8bc3a68def25a8b915f10c72a371d850596714f3fe535e43ce8d9e4dd)